core/controllers/story_viewer.py

```diff
@@ -35,20 +48,35 @@ def get(self, story_id):
 
         story = story_services.get_story_by_id(story_id)
 
-        completed_nodes = [completed_node.to_dict()
-                           for completed_node in
-                           story_services.get_completed_nodes_in_story(
-                               self.user_id, story_id)]
+        completed_node_ids = [
+            completed_node.id for completed_node in
+            story_services.get_completed_nodes_in_story(self.user_id, story_id)]
+
+        ordered_nodes_dict = [
+            node.to_dict() for node in story.story_contents.get_ordered_nodes()
+            if node.exploration_id]
+        for node in ordered_nodes_dict:
+            node['completed'] = False
+            if node['id'] in completed_node_ids:
+                node['completed'] = True
+
+        exp_ids = [
+            node['exploration_id'] for node in ordered_nodes_dict]
+        exp_summary_dicts = (
+            summary_services.get_displayable_exp_summary_dicts_matching_ids(
+                exp_ids, user=self.user))
+        exp_summaries_dict_map = {
```

Comment:
Hm, wouldn't exp_summary_dicts already be in order by exp_ids, which matches the order of ordered_nodes_dict?

Reply:
Yeah, done. Removed this dict and used that list directly.